CDI 4.0 Build Compatible Extensions

Azquelt opened the pull request:
Changes to support Build Compatible Extensions (BCEs)

Weld provides [LiteExtensionTranslator](https://github.com/weld/core/blob/master/weld-lite-extension-translator/src/main/java/org/jboss/weld/lite/extension/translator/LiteExtensionTranslator.java) to implement support for BCEs using a portable extension. However, we need to pass in the classloader and list of BCE classes to its constructor which we didn't previously have a way to do.

Changes in this PR:
* Move the instantiation of SPI extensions from `WebsphereCDIDeployment` into the `Archive` classes
* Add a service to provide extension archives
* Provide extra extension archives for the `LiteExtensionTranslator` if the application has BCEs and we're using CDI-4.0

The tests include a workaround for a permissions issue which I think requires fixes in the `LiteExtensionTranslator` itself. I've opened [WELD-2705](https://issues.redhat.com/browse/WELD-2705) for this.

Fixes #19912

Azquelt committed:
CDI FAT tests for Build Compatible Extensions

Tests for:
* BCE in WAR adding a type
* BCE in EAR adding a type
* BCE modifying a type
* BCE which fails to load or is the wrong type

Azquelt committed:
CDI: Have archives construct SPI extensions

Have the ExtensionArchive implemenation do the construction of any SPI
Extensions, rather than just providing the class.

This allows for custom extension archives which need parameters passed
into the constructor.

Azquelt committed:
CDI: add service to provide extension archives

Add a service which can be implemented by other bundles to add
additional extension archives. This will first be used to add extensions
to translate build compatible extensions at runtime.

Azquelt committed:
CDI: enable build compatible extensions

Register the translation extension for build compatible extensions

Azquelt commented:
#build

Azquelt commented on dev/io.openliberty.cdi.4.0.internal.weld/src/io/openliberty/cdi40/internal/weld/LiteExtensionProvider.java:
```diff
@@ -0,0 +1,92 @@
+        try {
+            Class<?> clazz = cl.loadClass(className);
+            return clazz.asSubclass(BuildCompatibleExtension.class);
+        } catch (ClassNotFoundException e) {
+            Tr.error(tc, "The {0} BuildCompatibleExtension class could not be loaded", className);
+            throw new CDIException("The {0} BuildCompatibleExtension class could not be loaded", e);
+        } catch (ClassCastException e) {
+            Tr.error(tc, "The {0} class is not a subclass of BuildCompatibleExtension", className);
+            throw new CDIException("The {0} class is not a subclass of BuildCompatibleExtension", e);
+        }
```
This needs NLS

Azquelt replied:
NLS now added.

tevans78 commented on dev/com.ibm.ws.cdi.extension.bce_fat/.classpath:
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
I would have named this project something like `io.openliberty.cdi.4.0.internal.bce_fat`

Azquelt replied:
Hmm, I was following the existing layout of FATs where they're broken up by the area they test. Even though this is a 4.0 specific test, it's a new area so it seemed sensible to create a bucket for that area, rather than create a 4.0 bucket.

tevans78 replied:
So far, all of the existing buckets that do not have a version number in the name are for CDI 1.2 and above. There are separate buckets for CDI 2.0 and CDI 3.0 specific tests. e.g. `io.openliberty.cdi.3.0.internal_fat`

Azquelt replied:
Yuck :(

Azquelt replied:
Renamed to `io.openliberty.cdi.4.0.internal_fat`

tevans78 commented on dev/com.ibm.ws.cdi.extension.bce_fat/bnd.bnd:
```diff
@@ -0,0 +1,28 @@
+#*******************************************************************************
+# Copyright (c) 2018, 2021 IBM Corporation and others.
```
2022

tevans78 commented on dev/com.ibm.ws.cdi.extension.bce_fat/bnd.bnd:
```diff
@@ -0,0 +1,28 @@
+test.project: true
+
+-buildpath: \
+	io.openliberty.jakarta.cdi.4.0;version=latest,\
+	io.openliberty.jakarta.servlet.5.0;version=latest,\
+	com.ibm.websphere.javaee.annotation.1.2;version=latest,\
+	com.ibm.websphere.javaee.servlet.3.1
```
Surely these should all be jakarta EE9/10 versions?
```suggestion
	io.openliberty.jakarta.cdi.4.0;version=latest,\
	io.openliberty.jakarta.servlet.5.0;version=latest,\
	io.openliberty.jakarta.annotation.2.0;version=latest
```

Azquelt replied:
I had problems with this as classes which use `FATServlet` wouldn't compile.

The Java EE versions of all of our test framework are automatically added to the classpath because it's a fat project and there's no way to remove them. If I manually add `com.ibm.ws.componenttest.2.0` then it does work with just the Jakarta versions of the API on the classpath.

I assume this means we end up with both versions of com.ibm.ws.componenttest on the classpath which isn't ideal but it does at least seem to compile (and at runtime it should be fine because we'll have the right version of the componenttest feature loaded).

tevans78 replied:
I think that manually adding `com.ibm.ws.componenttest.2.0` is the correct thing to do. I was assuming that the build scripts were smart enough not to add the other version as well.

tevans78 commented on dev/com.ibm.ws.cdi.extension.bce_fat/build.gradle:
```diff
@@ -0,0 +1,11 @@
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+apply from: '../wlp-gradle/subprojects/fat.gradle'
```
is this file needed?

Azquelt replied:
Yes. The tasks to build and run the tests are not present without it.

tevans78 replied:
are you sure? I think that stuff is added by default for a `_fat` project. e.g. `io.openliberty.cdi.3.0.internal_fat` does not have a build.gradle file

Azquelt replied:
I must be missing something else in that case. I'll have a poke around and see if I can work out what it is.

Azquelt replied:
Needed `fat.project: true` instead of `test.project: true` in the bnd file.

tevans78 commented on dev/com.ibm.ws.cdi.extension.bce_fat/fat/src/com/ibm/ws/cdi/extension/bce_fat/BuildCompatibleExtensionsTest.java:
```diff
@@ -0,0 +1,88 @@
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
+ *******************************************************************************/
+package com.ibm.ws.cdi.extension.bce_fat;
```
```suggestion
package io.openliberty.cdi.internal.extension.bce.fat;
```

tevans78 commented on dev/com.ibm.ws.cdi.extension.bce_fat/fat/src/com/ibm/ws/cdi/extension/bce_fat/BuildCompatibleExtensionsTest.java:
```diff
@@ -0,0 +1,88 @@
+import com.ibm.websphere.simplicity.ShrinkHelper.DeployOptions;
+import com.ibm.ws.cdi.extension.bce_fat.basic.app.BasicBCEExtension;
+import com.ibm.ws.cdi.extension.bce_fat.basic.app.BasicBCETestServlet;
+import com.ibm.ws.cdi.extension.bce_fat.ear.lib.EarBCEExtension;
```
If at some point someone adds some more tests then using application/package names like "basic" and "ear" might not be descriptive enough? However, the best I can think of at the moment is "earBCETest" which isn't all that much better :-(
```suggestion
import io.openliberty.cdi.internal.extension.bce.fat.earBCETest.lib.EarBCEExtension;
```

Azquelt replied:
`basicear` and `basicwar` would probably be ok for the moment. These are just tests to ensure that BCEs work at all within .wars and .ears, they're not testing any specific required functionality.

Azquelt replied:
Updated the package names to `io.openliberty.cdi40.internal.fat.bce.{basicwar, basicear}`

tevans78 commented on dev/com.ibm.ws.cdi.extension.bce_fat/fat/src/com/ibm/ws/cdi/extension/bce_fat/BuildCompatibleExtensionsTest.java:
```diff
@@ -0,0 +1,88 @@
+
+import com.ibm.websphere.simplicity.ShrinkHelper;
+import com.ibm.websphere.simplicity.ShrinkHelper.DeployOptions;
+import com.ibm.ws.cdi.extension.bce_fat.basic.app.BasicBCEExtension;
```
isn't the "app" subpackage redundant? the app is called "basic"

Azquelt replied:
Good catch. Previously the `basic` package had the test and the `basic.app` package had the test app contents.

When I added the EAR test, it made sense to use one test class for both tests to avoid restarting the server.

tevans78 commented on dev/com.ibm.ws.cdi.extension.bce_fat/fat/src/com/ibm/ws/cdi/extension/bce_fat/basic/app/TestBean.java:
```diff
@@ -0,0 +1,7 @@
+package com.ibm.ws.cdi.extension.bce_fat.basic.app;
```
missing copyright

jhanders34 commented on dev/com.ibm.ws.cdi.internal/src/com/ibm/ws/cdi/internal/interfaces/WebSphereBeanDeploymentArchive.java:
```diff
@@ -289,13 +290,13 @@
     /**
      * Get hold of the class names for extensions registered via the SPI
```
Maybe this description should change now that it doesn't return class names.

Azquelt replied:
Good point :+1:

jhanders34 commented on dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/WebSphereCDIDeploymentImpl.java:
```diff
@@ -499,35 +497,25 @@ public ServiceRegistry getServices() {
+                for (Supplier<Object> spiExtensionSupplier : spiExtensionSuppliers) {
+                    Object extension = spiExtensionSupplier.get();
+                    if (!(extension instanceof Extension)) {
+                        throw new IllegalArgumentException(extension.getClass().getCanonicalName()
```
If it requires it to be an Extension, could the signature of the methods be `Supplier<Extension>` instead of `Supplier<Object>`.  I suppose that would require this check to be done in ExtensionArchiveImpl instead.  It does make the method signatures that takes / return the Supplier so much better though.

Azquelt replied:
Ideally yes, but at the moment we have some code which returns a javax Extension and other code which returns a jakarta Extension so we can't do that until we can compile against the transformed version of this bundle.

tevans78 replied:
ah! this answers one of my previous questions too

jhanders34 commented on dev/com.ibm.ws.cdi.extension.bce_fat/fat/src/com/ibm/ws/cdi/extension/bce_fat/FATSuite.java:
```diff
@@ -0,0 +1,21 @@
+import org.junit.runners.Suite;
+import org.junit.runners.Suite.SuiteClasses;
+
+@RunWith(Suite.class)
```
Since this fat is only testing EE 10 function, it will need to have the minimum Java annotation configured with a value of 11.

Azquelt replied:
Thanks, didn't think of that

Azquelt replied:
Added `@MinimumJavaLevel(javaLevel = 11)` to the suite.

tevans78 commented on dev/com.ibm.ws.cdi.extension.bce_fat/bnd.bnd:
```diff
@@ -0,0 +1,28 @@
+src: \
+    fat/src,\
+    
+tested.features:
```
Since @jhanders34 mentioned Java 11, does this bnd also need
```
javac.source: 11
javac.target: 11

Require-Capability: osgi.ee; filter:="(&(osgi.ee=JavaSE)(version=11))"
```

Azquelt replied:
It might be allowed to have this part, though I'm not sure it's necessary.
```
javac.source: 11
javac.target: 11
```

I don't think it wants `Require-Capability` since it's not actually making a bundle.

jhanders34 replied:
Can do the `javac.source` and `javac.target`, but do not need to do the Require-Capability since this is a FAT project and it doesn't create a bundle.

tevans78 commented on dev/com.ibm.ws.cdi.extension.bce_fat/fat/src/com/ibm/ws/cdi/extension/bce_fat/basic/app/beans.xml:
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<beans xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+xmlns="https://jakarta.ee/xml/ns/jakartaee"
+xsi:schemaLocation="https://jakarta.ee/xml/ns/jakartaee https://jakarta.ee/xml/ns/jakartaee/beans_3_0.xsd"
+version="3.0"
```
```suggestion
xsi:schemaLocation="https://jakarta.ee/xml/ns/jakartaee https://jakarta.ee/xml/ns/jakartaee/beans_4_0.xsd"
version="4.0"
```
Not sure if this test _needs_ it to be version 4.0 but it would be a good idea to test it anyway

tevans78 commented on dev/com.ibm.ws.cdi.extension.bce_fat/fat/src/com/ibm/ws/cdi/extension/bce_fat/basic/app/permissions.xml:
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<permissions xmlns="http://xmlns.jcp.org/xml/ns/javaee"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee
+        http://xmlns.jcp.org/xml/ns/javaee/permissions_7.xsd"
+    version="7">
```
Similarly, I know this does not _need_ to be EE10 but would be useful to test
```suggestion
    xsi:schemaLocation="https://jakarta.ee/xml/ns/jakartaee
        https://jakarta.ee/xml/ns/jakartaee/permissions_10.xsd"
    version="10">
```

Azquelt replied:
Ah, I didn't realise there was an EE10 version of the permissions schema. The EE9.1 platform spec still links to the 7 version.

jhanders34 replied:
https://jakarta.ee/xml/ns/jakartaee/ shows you what they look like for each of the releases.

Azquelt replied:
Thanks, that's really helpful.

Azquelt replied:
It looks like we don't yet support version 10 of permissions.xml so I've set it to 9.

jhanders34 replied:
BTW, support is on its way.  Not sure if it will be here next week or after the GM build on March 2.

tevans78 commented on dev/com.ibm.ws.cdi.extension.bce_fat/fat/src/com/ibm/ws/cdi/extension/bce_fat/ear/lib/beans.xml:
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<beans xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xmlns="https://jakarta.ee/xml/ns/jakartaee"
+    xsi:schemaLocation="https://jakarta.ee/xml/ns/jakartaee https://jakarta.ee/xml/ns/jakartaee/beans_3_0.xsd"
```
version 4.0?

tevans78 commented on dev/com.ibm.ws.cdi.extension.bce_fat/fat/src/com/ibm/ws/cdi/extension/bce_fat/ear/lib/earPermissions.xml:
```diff
@@ -0,0 +1,12 @@
+<permissions xmlns="http://xmlns.jcp.org/xml/ns/javaee"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://xmlns.jcp.org/xml/ns/javaee
+        http://xmlns.jcp.org/xml/ns/javaee/permissions_7.xsd"
```
version 10?

tevans78 commented on dev/com.ibm.ws.cdi.internal/src/com/ibm/ws/cdi/internal/archive/liberty/ExtensionArchiveImpl.java:
```diff
@@ -61,7 +69,22 @@ public boolean isExtClassesOnly() {
     @Override
-    public Set<String> getSPIExtensionClasses() {
-        return spiExtensions;
+    public Set<Supplier<Object>> getSPIExtensionSuppliers() {
```
Using a Supplier seems like a good idea, so why not change the `getExtensionClasses` method too? Could be a follow up PR?

tevans78 replied:
hmm actually that looks hard :-(

Azquelt replied:
I'll have a closer look at this. I thought there was somewhere where we needed the actual extension classes, possibly because Weld was doing the instantiation based on the services files, but I'll double check that. It would be nicer to have them be consistent.

tevans78 replied:
Why isn't the generic `Set<Supplier<? extends Extension>>` ?

Azquelt replied:
Ideally it would be but at the moment we have some code which returns a javax Extension and other code which returns a jakarta Extension so we can't do that until we can compile against the transformed version of this bundle.

tevans78 commented on dev/com.ibm.ws.cdi.internal/src/com/ibm/ws/cdi/internal/archive/liberty/ExtensionArchiveImpl.java:
```diff
@@ -61,7 +69,22 @@ public boolean isExtClassesOnly() {
+        Set<Supplier<Object>> result = new HashSet<>();
+        for (String className : spiExtensions) {
+            try {
+                Class<?> clazz = getClassLoader().loadClass(className);
+                Constructor<?> constructor = clazz.getConstructor();
+                Object instance = constructor.newInstance();
```
the comment on the other method says that we are returning strings in order to delay the instantiation of the class. don't we want to do that here too? doesn't calling newInstance here defeat the point of using the supplier?

Azquelt replied:
Fair point, the class load and instantiation should possibly be inside the supplier. Will just have to check the error handling.

Azquelt replied:
On further investigation, I don't think this is necessary. This code is for creating extensions contributed via the CDIExtensionMetadata service where they're provided as `Class` objects anyway. There's no reason for this code to be doing a load by name, we could be passing through the classes to here, but even if we are loading by name we can be pretty sure it will work and work fast because the class is already loaded.

The previous method covers regular extensions as well, which are loaded by name after being found from the services file.

tevans78 replied:
So you're going to change the return type to `Set<Class<?>>`?

Azquelt replied:
No, we can't return classes because `LiteExtensionArchive` needs to be able to call a different constructor.

I'm a bit torn between just returning the extension object itself, or returning a supplier that calls the constructor. I guess it should be one or the other.

Azquelt replied:
We _could_ refactor the code so that we already have a `Set<Class<?>>` at this point, and produce the suppliers from that rather than from `Set<String>` but that's really outside the scope of this PR.

tevans78 commented on dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/ProbeExtensionArchive.java:
```diff
@@ -177,8 +178,8 @@ public boolean isExtClassesOnly() {
     @Override
-    public Set<String> getSPIExtensionClasses() {
-        return Collections.<String>emptySet();
+    public Set<Supplier<Object>> getSPIExtensionSuppliers() {
```
I know it is borderline but I think this is a functional change that should have a copyright update

tevans78 commented on dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/WebSphereCDIDeploymentImpl.java:
```diff
@@ -23,6 +23,7 @@
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.ExecutorService;
+import java.util.function.Supplier;
```
copyright date

Azquelt replied:
Thanks, not sure how I missed this one...

tevans78 commented on dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/liberty/CDIRuntimeImpl.java:
```diff
@@ -114,6 +116,9 @@
     @Reference(name = "beanParser", service = BeanParser.class)
     private BeanParser beanParser;
```
Remind me why I called this a "BeanParser"? It should have been "BeansXMLParser" :-(

Azquelt replied:
I agree, should be fine to change it as it's all internal though, right?

tevans78 commented on dev/io.openliberty.cdi.4.0.internal.weld/src/io/openliberty/cdi40/internal/weld/LiteExtensionArchive.java:
```diff
@@ -0,0 +1,181 @@
+    /** {@inheritDoc} */
+    @Override
+    public J2EEName getJ2EEName() throws CDIException {
+        return null;
```
Appears that returning null is the correct thing to do but maybe we need a comment to explain why?

tevans78 replied:
same comment for several of the other methods in this class

Azquelt replied:
Probably a good idea, I think I just copied from the `ProbeExtensionArchive`.

Azquelt replied:
For this particular method, the javadoc on `CDIArchive` says it can return `null` if there's no J2EE name, but for many of the others it doesn't, so I'm not sure if `null` is a valid return value or if the methods are just not called on these archives for some reason.

Does warrant some further investigation, I'll see if I can look at the existing code and update the `CDIArchive` javadoc.

tevans78 replied:
The handling of a null J2EEName doesn't seem 100% safe/consistent. It is a shame that we need it to be a J2EEName and not a more free form string ID.

Azquelt replied:
From what I can see, it's either null-checked or called on module archives.

tevans78 replied:
OK, this PR does not change that situation. So _if_ it is a problem we can deal with it separately.

Azquelt replied:
I've updated the Javadoc on CDIArchive based on what the existing code already does to indicate under what circumstances methods are allows to return `null`.

tevans78 commented on dev/io.openliberty.cdi.4.0.internal.weld/src/io/openliberty/cdi40/internal/weld/LiteExtensionArchive.java:
```diff
@@ -0,0 +1,181 @@
+    /** {@inheritDoc} */
+    @Override
+    public boolean applicationBDAsVisible() {
+        return false; // TODO: might need to be true?
```
is there a test to help decide this?

Azquelt replied:
I _suspect_ it will end up not needing to be `true` since I think extensions see everything anyway and this archive will always only contain a single extension.

Azquelt replied:
So any test which tried to enhance or inspect an application bean should cover this.

Azquelt replied:
Added an enhancement test and removed the TODO comment.

Azquelt commented:
I've renamed the FAT project and done the simple review changes as a first step, now I'll go through the ones which are less trivial or still have questions to answer.

Azquelt commented:
Personal build started manually, including running tests on Java 8: https://wasrtc.hursley.ibm.com:9443/jazz/web/projects/WS-CD#action=com.ibm.team.build.viewResult&id=_jB0VUIqREey7trZ3OKxvHw

Azquelt committed:
Update CDIArchive javadoc

Many of the methods on CDIArchive may and do return null in certain
circumstances but this wasn't documented.

Emily-Jiang commented on dev/io.openliberty.cdi.4.0.internal_fat/fat/src/io/openliberty/cdi40/internal/fat/bce/enhance/beans.xml:
```diff
@@ -0,0 +1,6 @@
+<beans xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xmlns="https://jakarta.ee/xml/ns/jakartaee"
+    xsi:schemaLocation="https://jakarta.ee/xml/ns/jakartaee https://jakarta.ee/xml/ns/jakartaee/beans_4_0.xsd"
+    version="4.0" bean-discovery-mode="annotated">
```
maybe make this beans.xml as empty one to cover the fact of empty beans.xml is the same as the one with `bean-discovery-mode="annotated"` specified?

Azquelt replied:
That makes sense, but I didn't think we'd implemented that yet.

I thought Tom had put the code in place to be able to change the behaviour of an empty beans.xml but had left it set to the old behaviour so that he could run existing FAT tests against it.

tevans78 replied:
The change in empty beans.xml behaviour will be tested along with the new config option... So slightly more value in making sure that `beans_4_0.xsd` works OK?

tevans78 replied:
The new configuration does not exist yet so it is currently hard-coded to `ALL`.

Emily-Jiang replied:
In CDI 4.0, by default, empty beans.xml should behave same as if annotated. It should not hard coded to `ALL` for cdi 4.0. For CDI 4.0, the default is `Annoated`. The new config can change the value to `All`.

Azquelt committed:
CDI: update test for SPI extension error

The error message here has been extended slightly as the code handling
the error might now deal with different failures.

Azquelt commented:
I think that's all the comments addressed.

#libby
#build

Azquelt commented:
Last build failed the code checkout
#build

dmuelle commented on dev/com.ibm.ws.cdi.internal/resources/com/ibm/ws/cdi/internal/resources/CDI.nlsprops:
```diff
@@ -76,3 +76,15 @@ spi.extension.failed.to.construct.CWOWB1010E.useraction=Fix the error that is pr
 spi.extension.class.in.different.bundle.CWOWB1011E.explanation=The class is ignored because it was registered with the getExtensions() or getBeanClasses() method of a CDIExtensionMetadata implementation and is in a different bundle than the CDIExtensionMetadata implementation.
 spi.extension.class.in.different.bundle.CWOWB1011E.useraction=Move the affected class or create a new CDIExtensionMetadata implementation in the correct bundle.
+
+exception.creating.extensions.CWOWB1012E=CWOWB1012E: Exception creating extensions for the {0} archive. The error was: {1}
```
suggested edit
Exception creating extensions for the {0} archive.
-->
CDI encountered an exception when it created extensions for the {0} archive.

If not "CDI", then substitute the entity that creates the extensions

dmuelle replied:
add a period after {1}

Azquelt replied:
I don't think we can safely put a period after `{1}`. The error message may itself be a translated string which could end in a period.

dmuelle commented on dev/com.ibm.ws.cdi.internal/resources/com/ibm/ws/cdi/internal/resources/CDI.nlsprops:
```diff
@@ -76,3 +76,15 @@ spi.extension.failed.to.construct.CWOWB1010E.useraction=Fix the error that is pr
 spi.extension.class.in.different.bundle.CWOWB1011E.useraction=Move the affected class or create a new CDIExtensionMetadata implementation in the correct bundle.
+
+exception.creating.extensions.CWOWB1012E=CWOWB1012E: Exception creating extensions for the {0} archive. The error was: {1}
+exception.creating.extensions.CWOWB1012E.explanation=CDI failed to load or invoke the construction of an extension class
```
add period here and in action message

dmuelle commented on dev/com.ibm.ws.cdi.internal/resources/com/ibm/ws/cdi/internal/resources/CDI.nlsprops:
```diff
@@ -76,3 +76,15 @@ spi.extension.failed.to.construct.CWOWB1010E.useraction=Fix the error that is pr
+exception.creating.extensions.CWOWB1012E.explanation=CDI failed to load or invoke the construction of an extension class
+exception.creating.extensions.CWOWB1012E.useraction=Fix the error that is preventing the extension from being created
+
+bce.not.loadable.CWOWB1013E=CWOWB1013E: The {0} BuildCompatibleExtension class cannot not be loaded
```
add period

dmuelle commented on dev/com.ibm.ws.cdi.internal/resources/com/ibm/ws/cdi/internal/resources/CDI.nlsprops:
```diff
@@ -76,3 +76,15 @@ spi.extension.failed.to.construct.CWOWB1010E.useraction=Fix the error that is pr
+exception.creating.extensions.CWOWB1012E.useraction=Fix the error that is preventing the extension from being created
+
+bce.not.loadable.CWOWB1013E=CWOWB1013E: The {0} BuildCompatibleExtension class cannot not be loaded
+bce.not.loadable.CWOWB1013E.explanation=The application includes a META-INF/services/jakarta.enterprise.inject.build.compatible.spi.BuildCompatibleExtension file which includes the class, but CDI could not load the class
```
file which includes the class,
-->
file that includes the class,

dmuelle commented on dev/com.ibm.ws.cdi.internal/resources/com/ibm/ws/cdi/internal/resources/CDI.nlsprops:
```diff
@@ -76,3 +76,15 @@ spi.extension.failed.to.construct.CWOWB1010E.useraction=Fix the error that is pr
+
+bce.not.loadable.CWOWB1013E=CWOWB1013E: The {0} BuildCompatibleExtension class cannot not be loaded
+bce.not.loadable.CWOWB1013E.explanation=The application includes a META-INF/services/jakarta.enterprise.inject.build.compatible.spi.BuildCompatibleExtension file which includes the class, but CDI could not load the class
+bce.not.loadable.CWOWB1013E.useraction=Ensure the class is present in the application and can be loaded. If the class is present in the application, there may be an FFDC file which has more information about why the class could not be loaded.
```
there may be an FFDC file which has more information..
--->
an FFDC file might have more information...

dmuelle commented on dev/com.ibm.ws.cdi.internal/resources/com/ibm/ws/cdi/internal/resources/CDI.nlsprops:
```diff
@@ -76,3 +76,15 @@ spi.extension.failed.to.construct.CWOWB1010E.useraction=Fix the error that is pr
+bce.not.loadable.CWOWB1013E.explanation=The application includes a META-INF/services/jakarta.enterprise.inject.build.compatible.spi.BuildCompatibleExtension file which includes the class, but CDI could not load the class
+bce.not.loadable.CWOWB1013E.useraction=Ensure the class is present in the application and can be loaded. If the class is present in the application, there may be an FFDC file which has more information about why the class could not be loaded.
+
+bce.does.not.implement.bce.CWOWB1014E=CWOWB1014E: The {0} class does not implement the BuildCompatibleExtension interface
```
add period

dmuelle commented on dev/com.ibm.ws.cdi.internal/resources/com/ibm/ws/cdi/internal/resources/CDI.nlsprops:
```diff
@@ -76,3 +76,15 @@ spi.extension.failed.to.construct.CWOWB1010E.useraction=Fix the error that is pr
+bce.not.loadable.CWOWB1013E.useraction=Ensure the class is present in the application and can be loaded. If the class is present in the application, there may be an FFDC file which has more information about why the class could not be loaded.
+
+bce.does.not.implement.bce.CWOWB1014E=CWOWB1014E: The {0} class does not implement the BuildCompatibleExtension interface
+bce.does.not.implement.bce.CWOWB1014E.explanation=The application includes a META-INF/services/jakarta.enterprise.inject.build.compatible.spi.BuildCompatibleExtension file which includes the class, but the class does not implement jakarta.enterprise.inject.build.compatible.spi.BuildCompatibleExtension
```
file which includes the class,
-->
file that includes the class,

dmuelle commented on dev/com.ibm.ws.cdi.internal/resources/com/ibm/ws/cdi/internal/resources/CDI.nlsprops:
```diff
@@ -76,3 +76,15 @@ spi.extension.failed.to.construct.CWOWB1010E.useraction=Fix the error that is pr
+
+bce.does.not.implement.bce.CWOWB1014E=CWOWB1014E: The {0} class does not implement the BuildCompatibleExtension interface
+bce.does.not.implement.bce.CWOWB1014E.explanation=The application includes a META-INF/services/jakarta.enterprise.inject.build.compatible.spi.BuildCompatibleExtension file which includes the class, but the class does not implement jakarta.enterprise.inject.build.compatible.spi.BuildCompatibleExtension
+bce.does.not.implement.bce.CWOWB1014E.useraction=Change the class so that it implements jakarta.enterprise.inject.build.compatible.spi.BuildCompatibleExtension
```
add period

dmuelle reviewed (changes requested):
ID review- a few suggestions, mainly to add missing periods to some of the messages. Let me know if any questions, thanks

Azquelt commented:
Thank you @dmuelle, I've made all your suggested changes except one where I've left a comment above.

Azquelt commented:
#libby
#build

Azquelt commented:
#build
#libby

dmuelle reviewed (approved):
LGTM

Azquelt commented:
@jhanders34 Could you add an approval (or further comments) for the changes you requested? Github won't let me merge until you do.

@tevans78 is going to do the full review and I won't merge until he has done so.

tevans78 commented on dev/com.ibm.ws.cdi.internal/src/com/ibm/ws/cdi/internal/interfaces/ExtensionArchiveProvider.java:
```diff
@@ -0,0 +1,33 @@
+     *
+     * @param cdiRuntime the CDI runtime
+     * @param deployment the application deployment
+     * @param application the application
```
there is no application param

Azquelt replied:
Good spot

tevans78 commented on dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/CDIContainerImpl.java:
```diff
@@ -614,6 +615,10 @@ public String getCurrentApplicationContextID() {
             }
         }
 
+        for (ExtensionArchiveProvider provider : cdiRuntime.getExtensionArchiveProviders()) {
```
most of the main sections in this method have short explainations about what they are doing ... maybe this needs a comment adding?

Azquelt replied:
Yes, I can do that.

tevans78 commented on dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/BDAFactory.java:
```diff
@@ -91,7 +91,7 @@ public static WebSphereBeanDeploymentArchive createBDA(WebSphereCDIDeployment de
                                                        eeModuleDescriptor);
 
-        bda.setSPIExtensionClassNames(extensionArchive.getSPIExtensionClasses());
+        bda.setSPIExtensionSuppliers(extensionArchive.getSPIExtensionSuppliers());
```
borderline but copyright date change??

Azquelt replied:
Sure

tevans78 commented on dev/io.openliberty.cdi.4.0.internal.weld/src/io/openliberty/cdi40/internal/weld/LiteExtensionProvider.java:
```diff
@@ -0,0 +1,92 @@
+/**
+ * Provides the weld-lite-extension-translator extension when the deployed application has build compatible extensions
+ */
+@Component(configurationPolicy = ConfigurationPolicy.IGNORE)
```
while not strictly required, would it be a good idea to set `service = ExtensionArchiveProvider.class`?

Azquelt replied:
I don't usually do that. Many of our small components do only want to expose the interfaces they implement as services, so it should be fairly common.

tevans78 commented on dev/io.openliberty.cdi.4.0.internal_fat/bnd.bnd:
```diff
@@ -0,0 +1,31 @@
+javac.source: 11
+javac.target: 11
+
+tested.features:
```
not even cdi-4.0?!

Azquelt replied:
I think `tested.features` doesn't need to list any features which are listed directly in the test server.xmls. It's mostly only needed when the bucket is repeated and so the test is run with features which aren't listed. The test framework will fail the server start if you're missing features from tested.features.

tevans78 commented on dev/io.openliberty.cdi.4.0.internal_fat/fat/src/io/openliberty/cdi40/internal/fat/bce/BuildCompatibleExtensionsTest.java:
```diff
@@ -0,0 +1,167 @@
+    }
+
+    @AfterClass
+    public static void teardown() throws Exception {
```
Would it be a good idea to separate the "good" tests from the "bad", put them on separate servers?

Azquelt replied:
So that we only allow errors on the "bad" server? Yeah, that sounds like a good idea.

It's annoying that we only check for errors when we shut down the server, rather than at the end of every test like we do for FFDCs. I wonder if we can change that.

tevans78 commented on dev/io.openliberty.cdi.4.0.internal_fat/fat/src/io/openliberty/cdi40/internal/fat/bce/basicwar/permissions.xml:
```diff
@@ -0,0 +1,13 @@
+    <!-- Update to version=10 when implemented -->
+    <permission>
+        <class-name>java.lang.RuntimePermission</class-name>
+        <name>accessDeclaredMembers</name>
```
Which bit of application code needs this permission?

Azquelt replied:
Ah, this should have a comment.

The application code doesn't need it, but it uses the CDI SPI, which allows access to declared members.

Ideally, CDI would declare which permissions are needed by which methods, or define and check its own permission but currently it doesn't. With `SecurityManager` deprecated in Java 17, I think it's unlikely that this will change.

Weld _could_ wrap everything in `doPriviliged` calls, but that would completely subvert these runtime security checks by allowing access to the same information by a different API, so I think it's better not to do that.

jhanders34 reviewed (approved):
I am marking it approved since what I commented on has been corrected to the point that it could be corrected.  @tevans78 is doing the more extensive review.

Azquelt commented:
Squashed review changes into the relevant commits and verified that the end result was identical, then rebased on the current integration.

#libby
#build
